openwisp_radius/api/views.py

```diff
@@ -70,9 +79,40 @@
 
 
 class TokenAuthentication(BaseAuthentication):
+    def _radius_token_authenticate(self, request):
+        # cached_orgid exists only for users authenticated
+        # successfully in past 24 hours
+        username = request.data.get('username')
+        cached_orgid = cache.get(f'rt-{username}')
+        if cached_orgid:
+            return (AnonymousUser(), cached_orgid)
+        else:
+            try:
+                radtoken = RadiusToken.objects.get(user__username=username)
+            except RadiusToken.DoesNotExist:
+                message = _(
+                    'Radius token does not exist. Obtain a new radius token '
+                    'or provide the organization UUID and API token.'
+                )
+                logger.warning(message)
+                raise NotAuthenticated(message)
+            if radtoken.can_auth:
+                org_uuid = str(radtoken.organization_id)
+                cache.set(f'rt-{username}', org_uuid, 86400)
+                return (AnonymousUser(), org_uuid)
```

Comment:
I think there's a misunderstanding here.

The `can_auth` attribute should be used when the user tries to log in via the captive page, which happens in the `authorize` method: https://github.com/openwisp/openwisp-radius/blob/master/openwisp_radius/api/views.py#L171-L183
I see you already changed that method in this PR so that should be enough.

If we leave this change here, after the user authenticates via the captive portal, their token will be flagged as `can_auth` False and if for some reason the cache is cleaned, this if block will be executed and the system won't be able to recognize accounting requests for this user (accounting requests are sent periodically by the NAS while the user is using the network).

We have to explicitly document this, `can_auth` indicates whether the token can be used for the freeradius authorization step (authorize view).

In this class instead we just deal with finding out to which organization the incoming API requests coming from freeradius belong.

It may be a good idea to change the name of this class altogether to avoid ambiguities.

We could rename it like:

- FreeradiusApiAuthentication
- FreeradiusAuthentication
- RadiusMultitenancyAuth

Or anything else that makes it explicit that here we're implementing multi-tenancy for HTTP API requests coming from freeradius, we're just making the system aware to which organization the request coming from freeradius belongs to, because freeradius itself has no concept of `organization` or `tenant`.